LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageProcessor.cs

```diff
@@ -141,27 +141,42 @@ private async Task<DownlinkPktFwdMessage> ProcessLoraMessage(LoRaMessageWrapper
                         dynamic fullPayload = JObject.FromObject(rxPk);
                         string jsonDataPayload = "";
                         uint fportUp = 0;
-                        bool isAckFromDevice = false;
+
                         if (loraMessage.LoRaPayloadMessage.GetLoRaMessage().Fport.Span.Length > 0)
                         {
                             fportUp = (uint)loraMessage.LoRaPayloadMessage.GetLoRaMessage().Fport.Span[0];
                         }
-                        else // this is an acknowledgment sent from the device
-                        {
-                            isAckFromDevice = true;
-                            fullPayload.deviceAck = true;
-                        }
+
                         fullPayload.port = fportUp;
                         fullPayload.fcnt = fcntup;
+                        List<KeyValuePair<String, String>> messageProperties = new List<KeyValuePair<String, String>>();
+                        fullPayload.eui = loraDeviceInfo.DevEUI;
+                        fullPayload.gatewayid = this.configuration.GatewayID;
+                        //Edge timestamp
+                        fullPayload.edgets = (long)((startTimeProcessing - new DateTime(1970, 1, 1)).TotalMilliseconds);
 
-                        // ACK from device or no decoder set in LoRa Device Twin. Simply return decryptedMessage
-                        if (isAckFromDevice || String.IsNullOrEmpty(loraDeviceInfo.SensorDecoder))
+                        //In case of ack message
+                        if ((loraMessage.LoRaPayloadMessage.GetLoRaMessage().Fctrl.Span[0] & (byte)FctrlEnum.Ack) == 32)
                         {
-                            if (String.IsNullOrEmpty(loraDeviceInfo.SensorDecoder))
+                            Logger.Log(loraDeviceInfo.DevEUI, String.Concat($"C2D Message ack received",
+                                loraDeviceInfo.LastConfirmedMessageId != "C2DMsgId" ? $" for message id {loraDeviceInfo.LastConfirmedMessageId}" : ""),
+                                Logger.LoggingLevel.Error);
+                            messageProperties.Add(new KeyValuePair<string, string>("C2DMsgConfirmed", loraDeviceInfo.LastConfirmedMessageId != "C2DMsgId" ?
+                            loraDeviceInfo.LastConfirmedMessageId : "C2D Msg Confirmation"));
+                            fullPayload.deviceAck = true;
+                            loraDeviceInfo.LastConfirmedMessageId = null;
+                            if(decryptedMessage== null)
                             {
-                                Logger.Log(loraDeviceInfo.DevEUI, $"no decoder set in device twin. port: {fportUp}", Logger.LoggingLevel.Full);
+                                string iotHubAckMsg = fullPayload.ToString(Newtonsoft.Json.Formatting.None);
+                                await loraDeviceInfo.HubSender.SendMessageAsync(iotHubAckMsg, messageProperties);
+                                return null;

```

Comment:
I think you cannot simply return, I think  it should continue or at least we should also call the update fcnt  _ = loraDeviceInfo.HubSender.UpdateFcntAsync(loraDeviceInfo.FCntUp, loraDeviceInfo.FCntDown); It is always dangerous to use return in the current logic